File: frontend/backups/purification_start/src/lib/schemas.ts
import { z } from 'zod';

/**
 * IDENTITY: SOTA Semantic Schemas
 * ROLE: Garantir integridade de dados entre Prisma, API e Motor Matemático.
 * PRINCIPLE: Economia Generalizada (Shannon) - Tipagem forte reduz entropia de erro.
 */

// Schema para Stacks (Array de números positivos)
export const StacksSchema = z.array( z.number().nonnegative() ).min( 2 );

// Schema para Prizes (Array de números positivos decrescentes)
export const PrizesSchema = z.array( z.number().nonnegative() );

// Schema para a Física da Mesa (PhysicsSnapshot)
export const PhysicsSnapshotSchema = z.object( {
    heroStack: z.number().positive(),
    villain1Stack: z.number().positive().optional(),
    villain2Stack: z.number().positive().optional(),
    pot: z.number().positive(),
    heroInvested: z.number().nonnegative(),
    edgeFactor: z.number().min( 0.5 ).max( 2 ).default( 1 ).optional(),
    position: z.enum( ['IP', 'OOP', 'BB', 'SB'] ),
    referenceStatus: z.enum( ['baseline', 'tilt', 'protecting', 'bubble'] ).default( 'baseline' ),
} );

// Schema para Requisição de Inferência (InferenceRequest)
export const InferenceRequestSchema = z.object( {
    prompt: z.string(),
    system_prompt: z.string().optional(),
    physics_snapshot: PhysicsSnapshotSchema.optional(),
    predictive_profile: z.record( z.string(), z.any() ).optional(),
    max_tokens: z.number().int().default( 1024 ),
    model: z.string().optional(),
} );

// Schema para entrada do Motor de Perspectiva
export const PerspectivaInputSchema = z.object( {
    stacks: StacksSchema,
    prizes: PrizesSchema,
    heroIdx: z.number().int().nonnegative(),
    villainIdx: z.number().int().nonnegative(),
    potSize: z.number().positive(),
    heroCost: z.number().nonnegative(),
    winProb: z.number().min( 0 ).max( 1 ),
    realizationFactor: z.number().min( 0 ).max( 2 ).default( 1 ),
    edgeBase: z.number().default( 1 ),
    numPlayersInPot: z.number().int().min( 2 ).default( 2 ),
    bountyValue: z.number().optional(),
    kappa: z.number().optional(),
    isNearPayjump: z.boolean().optional(),
    blindsRisingSoon: z.boolean().optional(),
    currentEquityPct: z.number().optional(),
    heroPosition: z.string().optional(),
    spr: z.number().optional(),
    investidoAcumulado: z.number().optional(),
    blindCost: z.number().optional(),
} );

// Schema para o resultado do Motor de Perspectiva (PerspectivaResult)
export const PerspectivaResultSchema = z.object( {
    handEquity: z.number(),
    currentEquityPct: z.number(),
    deltaWinPct: z.number(),
    deltaLosePct: z.number(),
    deltaFoldPct: z.number(),
    valuation: z.number(),
    rioLiability: z.number(),
    fgsHealth: z.number(),
    survivalPressure: z.number(),
    dynamicEvFold: z.number(),
    perspectivaPct: z.number(),
    amortizedEdge: z.number(),
    ci: z.number(),
    marginInstability: z.number(),
    threshEq: z.number(),
    realizationFactor: z.number(),
    isActionBetterThanFold: z.boolean(),
    diagnostico: z.string(),
    bountyPower: z.number(),
    currentMapaICM: z.array( z.number() ),
    winMapaICM: z.array( z.number() ),
    loseMapaICM: z.array( z.number() ),
} );

// === PERSPECTIVA SOTA v4.6 ===

export const PerspectiveMetricSchema = z.object( {
    scenarioId: z.string().optional(),
    baseState: z.record( z.string(), z.number() ).describe( "chipEvFold, icmValuation" ),
    dynamicModifiers: z.record( z.string(), z.number() ).describe( "timeToBlindJumpMinutes, payjumpProximityFactor, positionalUrgency" ),
    structuralLiabilities: z.record( z.string(), z.number() ).describe( "multiwayOpponents, reverseImpliedOddsPenalty" ),
    edgeRelative: z.record( z.string(), z.number() ).describe( "stackDepthBb, humanNoiseFactor, technicalSuperiority" ),
    insolvency: z.object( {
        potOddsRatio: z.number(),
        perspectiveUtility: z.number(),
        insolvencyCoefficient: z.number(),
        isViable: z.boolean(),
    } ),
} );

// === TELEMETRIA SOTA ===

export const TelemetryCategorySchema = z.enum( [
    'quiz', 'simulator', 'performance', 'error',
    'Risk Premium', 'Fundamentos SOTA', 'Bolha', 'Pós-Flop'
] as const );

export const TelemetryPayloadSchema = z.object( {
    category: TelemetryCategorySchema,
    componentName: z.string().nullable().optional(),
    scenarioContext: z.any().nullable().optional(),
    userAction: z.string().nullable().optional(),
    optimalAction: z.string().nullable().optional(),
    evLoss: z.number().default( 0 ),
    isCorrect: z.boolean().default( true ),
    latency: z.number().default( 0 ),
    metadata: z.record( z.string(), z.any() ).nullable().optional(),
} );

export type PhysicsSnapshot = z.infer<typeof PhysicsSnapshotSchema>;
export type InferenceRequest = z.infer<typeof InferenceRequestSchema>;
export type PerspectivaInputValidated = z.infer<typeof PerspectivaInputSchema>;
export type TelemetryPayload = z.input<typeof TelemetryPayloadSchema>;

File: core/schemas.py
# core/schemas.py
# ruff: noqa: N815
"""Esquemas base de Pydantic SOTA."""

from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

TaskMetadata = dict[str, Any]


class Task(BaseModel):
    """Modelo SOTA de Tarefa para a fila."""

    id: str
    description: str
    status: Literal["pending", "running", "completed", "failed", "cancelled"] = "pending"
    timestamp: str
    agent: str = Field(..., pattern=r"^@[\w]+$")
    completedAt: str | None = None
    model: str | None = None
    metadata: dict[str, Any] = Field(
        default_factory=dict,
        description=("Metadados da tarefa, incluindo 'observers' para notificacoes e telemetria SOTA."),
    )

    @field_validator("agent")
    @classmethod
    def validate_agent_existence(cls, v: str) -> str:
        """Valida a consistencia do agente no core. Late import para respeitar hot-reload."""
        from core.config import VALID_AGENTS as _LIVE_AGENTS  # noqa: PLC0415 # pylint: disable=import-outside-toplevel

        if v not in _LIVE_AGENTS:
            raise ValueError(f"Agente desconhecido: {v}")
        return v


class GeneralTelemetry(BaseModel):
    """Esquema SOTA para Telemetria Geral/UI/Quiz (Paridade Zod)."""

    category: Literal[
        "quiz",
        "simulator",
        "performance",
        "error",
        "Risk Premium",
        "Fundamentos SOTA",
        "Bolha",
        "Pos-Flop",
    ]
    componentName: str | None = None
    scenarioContext: dict[str, Any] | list[Any] | str | None = None
    userAction: str | None = None
    optimalAction: str | None = None
    evLoss: float = 0.0
    isCorrect: bool = True
    latency: float = 0.0
    metadata: dict[str, Any] | None = None

    @field_validator("category", mode="before")
    @classmethod
    def normalize_category(cls, v: str) -> str:
        """Normaliza categorias para o padrao SOTA (ASCII)."""
        mapping = {
            "Pós-Flop": "Pos-Flop",
            "Pos Flop": "Pos-Flop",
        }
        return mapping.get(v, v)


class SOTAMetrics(BaseModel):
    """Isomorfismo SOTA: Schema compartilhado para metrica de IA."""

    esperanca: float = Field(..., description="Vetor de ganho esperado (PM)")
    expectativa: float = Field(..., description="Vetor bruto de EV")
    perspectiva: float = Field(..., description="Expectativa ajustada por risco RIO")
    ci: float = Field(..., description="Coeficiente de Insolvencia")
    is_solvent: bool
    is_actionable: bool


class RAGQuery(BaseModel):
    """Schema para consultas RAG (Retrieval Augmented Generation)."""

    query: str
    top_k: int = 5
    threshold: float = 0.75
    metadata_filter: dict[str, Any] | None = None


class LLMConfig(BaseModel):
    """Configuracao de hardware para VRAM e RAM."""

    vram_limit_gb: int = 12
    ram_reserve_gb: int = 4
    gpu_layers: int = 32
    context_window: int = 8192


class PerspectivaResult(BaseModel):
    """Resultado unificado do Motor de Perspectiva (Paridade Zod)."""

    handEquity: float
    currentEquityPct: float
    deltaWinPct: float
    deltaLosePct: float
    deltaFoldPct: float
    valuation: float
    rioLiability: float
    fgsHealth: float
    survivalPressure: float
    dynamicEvFold: float
    perspectivaPct: float
    amortizedEdge: float
    ci: float
    marginInstability: float
    threshEq: float
    realizationFactor: float
    isActionBetterThanFold: bool
    diagnostico: str
    bountyPower: float
    currentMapaICM: list[float]
    winMapaICM: list[float]
    loseMapaICM: list[float]


class InsolvencyMetrics(BaseModel):
    """Estrutura detalhada de insolvencia."""

    potOddsRatio: float
    perspectiveUtility: float
    insolvencyCoefficient: float | None = None
    isViable: bool


class PerspectiveMetric(BaseModel):
    """Esquema SOTA para Metricas de Perspectiva (Paridade Zod)."""

    scenarioId: str | None = None
    baseState: dict[str, float] = Field(..., description="chipEvFold, icmValuation")
    dynamicModifiers: dict[str, float] = Field(
        ...,
        description="timeToBlindJumpMinutes, payjumpProximityFactor, positionalUrgency",
    )
    structuralLiabilities: dict[str, float] = Field(..., description="multiwayOpponents, reverseImpliedOddsPenalty")
    edgeRelative: dict[str, float] = Field(..., description="stackDepthBb, humanNoiseFactor, technicalSuperiority")
    insolvency: InsolvencyMetrics

    def flatten(self) -> dict[str, Any]:
        """Achata a estrutura para persistencia no SQLite/CSV."""
        return {
            "scenarioId": self.scenarioId,
            **self.baseState,
            **self.dynamicModifiers,
            **self.structuralLiabilities,
            **self.edgeRelative,
            **self.insolvency.model_dump(),
        }
